Add boss HP% to overlay

jiawei-zheng-0 opened the pull request:
![electron_2022-12-27_16-51-50](https://user-images.githubusercontent.com/16696564/209728025-88a52bca-a90e-4924-a43d-a69662ac63f2.png)

Adds boss current hp, max hp, and hp percentage to overlay.

Boss is determined by looking at sessionState's entities property and selecting the entity with the newest(highest) lastUpdate value and has a name that appears in encounters.js. Hence only guardians and raid boss hp would be shown.

I have tested change in guardian raids but won't be able to test in legion raids until after reset.

There are room for improvement and some bugs I saw.
1. Add new setting to toggle showing of boss hp, just like total dmg, total dps, and total tank.
2. When doing non-boss content the boss hp of last fought boss will appear. This is because sessionState's entities property will  store its entities for 10 minutes. In my opinion, sessionState should drop all entities upon zone change but I don't know a way other than changing loa-details-log-parser.
3. Boss HP will sometimes fail to load. This happens when loa-details-log-parser does not parse the name of the boss properly. I notice this very often when meter is started in middle of raid rather than before the raid starts. loa-details-log-parser would declare the name of entities to be hex values rather than their actual names. This is the same bug when player names appear as hex.

jiawei-zheng-0 committed:
Add Boss HP % to overlay

jiawei-zheng-0 committed:
Move sessionBoss logic to seperate func to allow early exit when boss is determined.

jiawei-zheng-0 committed:
Merge branch 'main' of https://github.com/jiaster/loa-details into feature-add-boss-hp%-overlay

Herysia commented:
First of all, thx for your contribution,
Regarding the few points you mentionned, i think this PR needs to take care of it
1> Yh adding as settings for it is required in my opinion
2> loa-details-log-parser does wipe it on zone change (initEnv), but keep it on phase transition( boss dead, wipe, ...). Are you sure it is not ghosting for another reason ?
3 > This is not rlly a bug, it's missing data, the game dispatch Npc names on spawn (usually when you load inside guardian), so if you start the meter after, you don't have names. We could add guessing but it's a lot of work for something that could be fixed just by starting the meter before changing zone (which anw is the best way to get a full run)

As an extra note, i think that writing the boss name like "Deskaluda: " instead of "Boss HP:" can be great, as well as an extra digit to health %

jiawei-zheng-0 committed:
Add Boss HP display to settings.
Boss Hp will dynaically change to name of Boss.

jiawei-zheng-0 committed:
Remove Math.floor for calculating Boss HP, oops

jiawei-zheng-0 commented:
I added Boss HP display to settings with it defaulting to disabled. It was much simpler than I imagined.
The boss text on overlay will now dynamically change to boss name, also added one digit after decimal.

For point 3, I guess I will ignore. 

For point 2, I am pretty sure entities are not being cleared out on zone change.
![electron_2022-12-27_18-29-56](https://user-images.githubusercontent.com/16696564/209736848-6daa479b-1b0a-405d-a19e-81fb5af04025.png)
That is from being in town->guardian raid->town. All the entities persist. I based my 10 minute expiration assumption off this https://github.com/karaeren/loa-details-log-parser/blob/b5184e5e9ffcf3c02d39b6c4bfa3634dc176cdc5/src/parser.ts#L203

Herysia commented:
Yh i see what you mean for 3>, i thought that it was an hard reset as for log parsing, but Eren somehow used soft reset onInitEnv instead. There is no reasons for it to be like that, i'll change it. Also, we use a fork of loa-details-log-parser that currently is in my personal repos, not many changes, but it's great to know

Edit: thx for pointing 3> out, it's probably the source of missing localplayer on live log sometimes (when the localplayer has been afk for 10mn before changing zone, then it'll be wiped and not registered as "isPlayer=true", which makes him invisible

jiawei-zheng-0 committed:
Merge branch 'lost-ark-dev:main' into feature-add-boss-hp%-overlay
